core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -252,129 +462,260 @@ public void consume(Consumer<Ranges> rangesConsumer, Consumer<DiscreteValues> va
     {
-        SortedRangeSet otherRangeSet = checkCompatibility(other);
+        SortedRangeSet that = checkCompatibility(other);
 
-        Iterator<Range> iterator1 = lowIndexedRanges.values().iterator();
-        Iterator<Range> iterator2 = otherRangeSet.lowIndexedRanges.values().iterator();
+        if (this.isNone() || that.isNone()) {
+            return false;
+        }
 
-        if (iterator1.hasNext() && iterator2.hasNext()) {
-            Range range1 = iterator1.next();
-            Range range2 = iterator2.next();
+        int thisRangeCount = this.getRangeCount();
+        int thatRangeCount = that.getRangeCount();
 
-            while (true) {
-                if (range1.overlaps(range2)) {
-                    return true;
-                }
-
-                if (range1.getHigh().compareTo(range2.getHigh()) <= 0) {
-                    if (!iterator1.hasNext()) {
-                        break;
-                    }
-                    range1 = iterator1.next();
-                }
-                else {
-                    if (!iterator2.hasNext()) {
-                        break;
-                    }
-                    range2 = iterator2.next();
-                }
+        int thisNextRangeIndex = 0;
+        int thatNextRangeIndex = 0;
+        while (thisNextRangeIndex < thisRangeCount && thatNextRangeIndex < thatRangeCount) {
+            RangeView thisCurrent = this.getRangeView(thisNextRangeIndex);
+            RangeView thatCurrent = that.getRangeView(thatNextRangeIndex);
+            if (thisCurrent.overlaps(thatCurrent)) {
+                return true;
+            }
+            int compare = thisCurrent.compareTo(thatCurrent);
+            if (compare < 0) {
+                thisNextRangeIndex++;
+            }
+            if (compare > 0) {
+                thatNextRangeIndex++;
             }
         }
 
         return false;
     }
 
     @Override
-    public SortedRangeSet union(ValueSet other)
+    public SortedRangeSet union(Collection<ValueSet> valueSets)
     {
-        SortedRangeSet otherRangeSet = checkCompatibility(other);
-        return new Builder(type)
-                .addAll(this.lowIndexedRanges.values())
-                .addAll(otherRangeSet.lowIndexedRanges.values())
-                .build();
+        if (this.isAll()) {
+            return this;
+        }
+
+        // Logically this organizes all value sets (this and valueSets) into a binary tree and merges them pairwise, bottom-up.
+        // TODO generalize union(SortedRangeSet) to merge multiple sources at once
+
+        List<SortedRangeSet> toUnion = new ArrayList<>(1 + valueSets.size());
+        toUnion.add(this);
+        for (ValueSet valueSet : valueSets) {
+            SortedRangeSet other = checkCompatibility(valueSet);
+            if (valueSet.isAll()) {
+                return other;
+            }
+            toUnion.add(other);
+        }
+
+        while (toUnion.size() > 1) {
```

Comment:
```java
        SortedRangeSet result = toUnion.get(0);
        for (int i = 1; i < toUnion.size(); ++i) {
            SortedRangeSet other = toUnion.get(i);
            result = result.union(other);
        }
        return result;
```

Is simpler (should work though I did not check). Is it worse?

Reply:
Also it may (or may not) make sense to check if the result is not `all` on each iteration and break sooner if further iteration is not needed.

Reply:
@losipiuk  wouldn't that be quadratic?

Reply:
True. I did not think of this one. Number of calls to `union()` would be the same. But number of operations on individual ranges is O(nlogn) in your case. Also yours is easier to parallelize.